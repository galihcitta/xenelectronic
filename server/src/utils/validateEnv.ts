import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
