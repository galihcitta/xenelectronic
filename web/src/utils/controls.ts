import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
