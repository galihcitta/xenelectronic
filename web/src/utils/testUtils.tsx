import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
