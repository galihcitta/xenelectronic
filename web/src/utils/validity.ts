import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
