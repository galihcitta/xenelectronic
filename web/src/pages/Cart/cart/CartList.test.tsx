import { render } from '@testing-library/react';
import { CartListProps, _CartList } from './CartList.component';
import { withProvider } from '../../../utils/testUtils';

const CartListWithProvider = withProvider(_CartList);

describe('<CartList />', () => {
  const props: CartListProps = {
    cartItems: [
      {
        character: 'modern',
        collection: 'clocks',
        color: 'black',
        description: 'lorem ipsum',
        id: '1',
        images: ['aa', 'bb'],
        name: 'Clock',
        price: '200',
        type: 'clocks'
      },
      {
        character: 'modern',
        collection: 'clocks',
        color: 'black',
        description: 'lorem ipsum',
        id: '2',
        images: ['aa', 'bb'],
        name: 'Clock',
        price: '200',
        type: 'clocks'
      }
    ]
  };

  const setup = (props: CartListProps) => {
    const utils = render(CartListWithProvider(props));
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('contains 2 items on the list', () => {
    const { getAllByText } = setup(props);
    expect(getAllByText(/clock/i)).toHaveLength(2);
  });

  it('contains warning message', () => {
    const { getByText } = setup({ ...props, cartItems: [] });
    expect(getByText(/cart is empty/i)).toBeInTheDocument();
  });
});
