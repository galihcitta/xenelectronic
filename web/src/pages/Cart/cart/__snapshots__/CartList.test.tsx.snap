// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartList /> renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="sc-bdVaJa UoSJW"
  >
    <li
      class="sc-htpNat jxfDOc"
    >
      <figure
        class="sc-bxivhb jItGcD"
      >
        <img
          alt="Clock"
          class="sc-ifAKCX fDAxYi"
          src="aa"
        />
      </figure>
      <div
        class="sc-EHOje OptIa"
      >
        <p
          class="sc-bZQynM fkXBRJ"
        >
          <strong>
            Clock
          </strong>
        </p>
        <p
          class="sc-bZQynM fkXBRJ"
        >
          <strong>
            $ 200
          </strong>
        </p>
        <p
          class="sc-bZQynM fkXBRJ"
        >
          Color: black
        </p>
        <div
          class="sc-gzVnrw cjfsfd"
        >
          <button
            class="sc-htoDjs kamfNv"
            data-testid="remove-btn"
          >
            -
          </button>
          <button
            class="sc-htoDjs kamfNv"
            data-testid="add-btn"
          >
            +
          </button>
        </div>
      </div>
      <button
        class="sc-dnqmqq eegCJY"
        data-testid="clear-btn"
      >
        x
      </button>
    </li>
    <li
      class="sc-htpNat jxfDOc"
    >
      <figure
        class="sc-bxivhb jItGcD"
      >
        <img
          alt="Clock"
          class="sc-ifAKCX fDAxYi"
          src="aa"
        />
      </figure>
      <div
        class="sc-EHOje OptIa"
      >
        <p
          class="sc-bZQynM fkXBRJ"
        >
          <strong>
            Clock
          </strong>
        </p>
        <p
          class="sc-bZQynM fkXBRJ"
        >
          <strong>
            $ 200
          </strong>
        </p>
        <p
          class="sc-bZQynM fkXBRJ"
        >
          Color: black
        </p>
        <div
          class="sc-gzVnrw cjfsfd"
        >
          <button
            class="sc-htoDjs kamfNv"
            data-testid="remove-btn"
          >
            -
          </button>
          <button
            class="sc-htoDjs kamfNv"
            data-testid="add-btn"
          >
            +
          </button>
        </div>
      </div>
      <button
        class="sc-dnqmqq eegCJY"
        data-testid="clear-btn"
      >
        x
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
